### PR TITLE
feat(tts): add SSML tag support for phoneme, sub, and say-as

### PR DESCRIPTION
**Pull request by smdesai**
### Why is this change needed?
Text-to-speech applications often require fine-grained control over pronunciation that automatic phoneme generation cannot provide. Names, acronyms, technical terms, and numbers frequently need explicit pronunciation guidance or special interpretation rules. This change adds SSML (Speech Synthesis Markup Language) tag support to Kokoro TTS, enabling developers to override pronunciations with IPA phonemes via <phoneme>, substitute text before synthesis with \<sub\>, and control how content types like dates, numbers, and telephone numbers are spoken using \<say-as>\. The implementation follows W3C SSML standards and is designed to co-exist with the existing markdown-style [word](/phonemes/) phonetic override syntax as well as the [word](replacement) syntax —both approaches can be used together in the same text, with phonetic overrides from both sources merged during preprocessing. This gives developers flexibility to choose whichever annotation style fits their use case while maintaining full backward compatibility.

### CODE REVIEW

File: Documentation/TTS/SSML.md [hunk 1]
```diff
@@ -0,0 +1,235 @@
+# SSML Tag Support
```
Comment (Alex-Wengg) [inline Documentation/TTS/SSML.md:1]: might be useful to include this link 
https://www.ipachart.com/#google_vignette

File: Documentation/TTS/SSML.md
Comment (Alex-Wengg) [inline Documentation/TTS/SSML.md:1]: any possibly edge cases like Invalid IPA that can't map to model vocabulary

Comment (Alex-Wengg) [inline Documentation/TTS/SSML.md:1]: https://www.w3.org/TR/speech-synthesis11/

what about 
  - `<break>` - pauses
  - `<prosody>` - rate, pitch, volume control
  - `<emphasis>` - stress levels
  - `<voice>` - voice switching mid-text

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        let sortedOverrides = phoneticOverrides.sorted { $0.wordIndex < $1.wordIndex }
+
+        return SSMLProcessingResult(text: workingText, phoneticOverrides: sortedOverrides)
+    }
+
+    // MARK: - Private Helpers
+
+    /// Count completed words before a given index in the text
+    /// This matches the word counting logic in TtsTextPreprocessor
+    private static func countWordsBeforeIndex(in text: String, index: String.Index) -> Int {
```
Comment (Alex-Wengg) [inline Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift:70]: whats the difference between SSMLProcessor.countWordsBeforeIndex and TtsTextPreprocessor.WordAccumulator. seem to do almost the same thing

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        var workingText = text
+        var phoneticOverrides: [TtsPhoneticOverride] = []
+
+        // Parse all SSML tags (returned in reverse document order for safe replacement)
+        let tags = SSMLTagParser.parse(text)
+
+        // Process tags in reverse order (from end to start) to preserve indices
+        for tag in tags {
+            switch tag.type {
+            case .phoneme(let alphabet, let ph, let content):
```
Comment (Alex-Wengg) [inline Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift]: is this used ?

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+        // If we ended in a word, it's a partial word - don't count it
+        // The word index points to the word that will contain our phoneme
+        return wordCount
+    }
+
+    /// Check if a character is an emoji
+    private static func isEmoji(_ char: Character) -> Bool {
+        guard let scalar = char.unicodeScalars.first else { return false }
+        return scalar.properties.isEmoji && scalar.properties.isEmojiPresentation
```
Comment (Alex-Wengg) [inline Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift]: don't we have this function in ttstextpreprocessor.swift  
https://github.com/FluidInference/FluidAudio/blob/b3ce777b43ee3b71cd9ad3846d7d7d8a6cd59103/Sources/FluidAudioTTS/TextToSpeech/Kokoro/Pipeline/Preprocess/TtsTextPreprocessor.swift#L648

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+    /// Count completed words before a given index in the text
+    /// This matches the word counting logic in TtsTextPreprocessor
+    private static func countWordsBeforeIndex(in text: String, index: String.Index) -> Int {
+        let prefix = String(text[..<index])
+        var wordCount = 0
+        var inWord = false
+
+        for char in prefix {
+            let isWordChar = char.isLetter || char.isNumber || char == "'" || isEmoji(char)
```
Comment (Alex-Wengg) [inline Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift]: what about apostrophe variants https://github.com/FluidInference/FluidAudio/blob/b3ce777b43ee3b71cd9ad3846d7d7d8a6cd59103/Sources/FluidAudioTTS/TextToSpeech/Kokoro/Pipeline/Preprocess/TtsTextPreprocessor.swift#L666

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLTagParser.swift [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+/// Regex-based parser for SSML tags
+/// Supports: <phoneme>, <sub>, <say-as>
+enum SSMLTagParser {
+
+    // MARK: - Regex Patterns
+
+    /// Pattern for <phoneme alphabet="ipa" ph="...">content</phoneme>
+    /// Captures: (1) alphabet (optional), (2) ph (required), (3) content
+    private static let phonemePattern = try! NSRegularExpression(
+        pattern: #"<phoneme(?:\s+alphabet\s*=\s*["']([^"']*)["'])?\s+ph\s*=\s*["']([^"']+)["']\s*>([^<]*)</phoneme>"#,
```
Comment (Alex-Wengg) [inline Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLTagParser.swift]: The regex expects attributes in a specific order:
```

  #"<phoneme(?:\s+alphabet\s*=\s*["']...)?\s+ph\s*=\s*["']..."#
  //          ^^^^^^^^^ alphabet first    ^^^^ then ph
```

  But XML/SSML allows any attribute order. These are all valid:
 ```
 <phoneme ph="x" alphabet="ipa">word</phoneme>     <!-- Won't match -->
  <phoneme alphabet="ipa" ph="x">word</phoneme>    <!-- Will match -->
```

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLTagParser.swift [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        return tags.sorted { $0.range.lowerBound > $1.range.lowerBound }
+    }
+
+    // MARK: - Private Parsing Methods
+
+    private static func parsePhoneme(_ text: String, nsText: NSString) -> [SSMLParsedTag] {
+        var tags: [SSMLParsedTag] = []
+        let range = NSRange(location: 0, length: nsText.length)
+
+        phonemePattern.enumerateMatches(in: text, options: [], range: range) { match, _, _ in
```
Comment (Alex-Wengg) [inline Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLTagParser.swift]: why not create one enumerateMatches function instead of these 3 separate class functions .

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SayAsInterpreter.swift [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+        guard components.count >= 2 else { return content }
+
+        let hours = components[0]
+        let minutes = components[1]
+
+        if minutes == 0 {
+            return "\(interpretCardinal(String(hours))) o'clock"
+        }
+
+        return "\(interpretCardinal(String(hours))) \(interpretCardinal(String(minutes)))"
```
Comment (Alex-Wengg) [inline Sources/FluidAudioTTS/TextToSpeech/SSML/SayAsInterpreter.swift:367]: would something like 
3:05 → "three five" instead of "three oh five"

occur

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SayAsInterpreter.swift [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+                }
+                return interpretCardinal(String(century)) + " hundred"
+            } else if year >= 2000 && year <= 2009 {
+                // 2001-2009 -> "two thousand one" etc.
+                return "two thousand " + interpretCardinal(String(remainder))
+            } else {
+                // 1985 -> "nineteen eighty five", 2024 -> "twenty twenty four"
+                let centuryPart = interpretCardinal(String(century))
+                let remainderPart = interpretCardinal(String(remainder))
+                return "\(centuryPart) \(remainderPart)"
```
Comment (Alex-Wengg) [inline Sources/FluidAudioTTS/TextToSpeech/SSML/SayAsInterpreter.swift:316]: could this occur . 
1905 → "nineteen five" instead of "nineteen oh five"

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SayAsInterpreter.swift
Comment (Alex-Wengg) [inline Sources/FluidAudioTTS/TextToSpeech/SSML/SayAsInterpreter.swift:1]: Duplicate code with TtsTextPreprocessor

  Both files have:
  - Digit words array
  - Decimal-to-word conversion
  - Time formatting with "oh"

File: Tests/FluidAudioTests/SSMLTests.swift
Comment (Alex-Wengg) [inline Tests/FluidAudioTests/SSMLTests.swift:1]: Additional Edge Cases that might worth considering

  - 3:05 → "three oh five" (missing "oh" bug)
  - 1905 → "nineteen oh five" (missing "oh" bug)
  - <phoneme ph="x" alphabet="ipa"> - reversed attribute order
  - Unicode content: "café", emoji, "日本"
  - Fractions: 1/3, 2/2, 3/100
  - SSML after $10 - word index drift after expansion
  - Adjacent tags: <phoneme>one</phoneme><phoneme>two</phoneme>
  - Ordinals: 111, 1000
  - Invalid date bounds: month 13, day 32
  - Whitespace/newlines in content

File: Documentation/TTS/SSML.md
Comment (smdesai) [reply Documentation/TTS/SSML.md:1]: The intention was to support the SSML tags <say-as>, <sub> and <phoneme> for now  and perhaps add these later.

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift (hunk 2 above)
Comment (smdesai) [reply Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift:70]: They're similar because they need to count words the same way for wordIndex to be consistent, but WordAccumulator does more work because it's also constructing the processed text during the markdown [word](/phonemes/) parsing pass.

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift (hunk 3 above)
Comment (smdesai) [reply Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift]: yes and changed to case .phoneme(_, let ph, let content):

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLTagParser.swift (hunk 7 above)
Comment (smdesai) [reply Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLTagParser.swift]: Refactored from 3 separate functions to one shared enumerateMatches function

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SayAsInterpreter.swift (hunk 8 above)
Comment (smdesai) [reply Sources/FluidAudioTTS/TextToSpeech/SSML/SayAsInterpreter.swift:367]: Yes it would, it's fixed and a test has been added.

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SayAsInterpreter.swift (hunk 9 above)
Comment (smdesai) [reply Sources/FluidAudioTTS/TextToSpeech/SSML/SayAsInterpreter.swift:316]: It's fixed and a test has been added.

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SayAsInterpreter.swift
Comment (smdesai) [reply Sources/FluidAudioTTS/TextToSpeech/SSML/SayAsInterpreter.swift:1]: Refactored to remove duplicate code.

File: Tests/FluidAudioTests/SSMLTests.swift
Comment (smdesai) [reply Tests/FluidAudioTests/SSMLTests.swift:1]: Good point. These have been added. Invalid date is interesting as Polly does not attempt to "fix" invalid SSML; it simply rejects the request, as all SSML input must be well-formed and adhere to the W3C SSML Specification and Amazon's specific supported tags. 

The approach here is a graceful fallback where the text is spoken but not interpreted. Thoughts?

File: Documentation/TTS/SSML.md
Comment (smdesai) [reply Documentation/TTS/SSML.md:1]: Invalid IPA is logged as a warning and falls back to normal pronunciation (lexicon lookup or eSpeak). The word is spoken just not with custom phonemes.

File: Tests/FluidAudioTests/SSMLTests.swift
Comment (Alex-Wengg) [reply Tests/FluidAudioTests/SSMLTests.swift:1]: that makes the most sense, always useful to log warnings too in these cases . or return a value to signal fallback

File: Documentation/TTS/SSML.md (hunk 1 above)
Comment (smdesai) [reply Documentation/TTS/SSML.md:1]: Link added as reference

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift (hunk 2 above)
Comment (Alex-Wengg) [reply Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift:70]: what do you think about  Pulling the core word-counting into a shared utility that both use. WordAccumulator can still do its text construction on top

**Comment by Alex-Wengg** [comment]
@smdesai please lmk when you are ready for another review

File: Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift (hunk 2 above)
Comment (smdesai) [reply Sources/FluidAudioTTS/TextToSpeech/SSML/SSMLProcessor.swift:70]: It makes sense, I've moved out the code so isWordCharacter is now shared

**Comment by smdesai** [comment]
@Alex-Wengg I've moved the code to a shared utility

**Review by Alex-Wengg — APPROVED** [review]

**Comment by Alex-Wengg** [comment]
@smdesai lmk if you can merge now that i have approved the pr

**Comment by smdesai** [comment]
@Alex-Wengg Yes I can merge and just did.
